Simplify Environment Patching

joefitzgerald opened the pull request:
This PR simplifies the logic used to determine when to patch the environment for users of linux and macOS. Instead of using a whitelist of shells and trying to detect how Atom was launched with environment variable markers, we simply modify `atom.sh` to set a environment variable that is only set when launched from the shell. If this environment variable is not detected, the environment is patched with the environment from a new, interactive, login shell.
- [x] Stop whitelisting shells
- [x] Introduce `ATOM_SUPPRESS_ENV_PATCHING` environment variable, which is only set in `atom.sh`
- [x] Manually test on Ubuntu
- [x] Manually test on macOS

Fixes #11910
Fixes #12535

joefitzgerald committed:
Get Environment From Shell On Linux

joefitzgerald committed:
Use ATOM_SUPPRESS_ENV_PATCHING Environment Variable

- Stop using shell whitelist

joefitzgerald committed:
Prevent Regression For Common Shells

- Fix lint warnings

thomasjo commented:
LGTM™ 🚀

thomasjo commented:
Let's get this in before rolling the 🚃 🚃 ? /cc @atom/feedback @nathansobo @lee-dohm

nathansobo commented on src/update-process-env.js:
```diff
@@ -36,6 +33,10 @@ function updateProcessEnv (launchEnv) {
       if (!ENVIRONMENT_VARIABLES_TO_PRESERVE.has(key)) {
         process.env[key] = envToAssign[key]
+      } else {
+        if (!process.env[key] && envToAssign[key]) {
```
So we'll always assign "preserved" environment variables if they're not already defined?

nathansobo replied:
Seems like this duplicated assignment could be eliminated by making the conditional:

```
if (!ENVIRONMENT_VARIABLES_TO_PRESERVE.has(key) || !process.env[key]) {
  process.env[key] = envToAssign[key]
}

```

joefitzgerald replied:
This is related to https://github.com/atom/atom/pull/12562#discussion-diff-78041627R91 but I think this new behavior may be too pessimistic. It would prevent the changing of an existing value in process.env which is not the expected behavior. I will write up a spec to validate this assumption and fix if true.

nathansobo commented:
If I'm reading this correctly, it seems like the name of the variable `ATOM_SUPPRESS_ENV_PATCHING` is a bit misleading. We will update the `process.env` of an existing instance of Atom regardless of this flag. The flag indicates whether we should attempt to invoke the user's shell to retrieve the environment... am I right about that?

nathansobo commented:
What about an environment variable like `ATOM_LAUNCHED_FROM_SHELL`. That seems like it would more clearly communicate the intent unless I'm missed something.

nathansobo commented on spec/update-process-env-spec.js:
```diff
@@ -124,30 +182,42 @@ describe('updateProcessEnv(launchEnv)', function () {
+        expect(shouldGetEnvFromShell({SHELL: '/usr/local/bin/fish'})).toBe(true)
       })
 
       it('returns false when the shell should not be patched', function () {
```
Maybe this should be phrased `returns false when the environment indicates that Atom was launched from a shell`

nathansobo commented on spec/update-process-env-spec.js:
```diff
@@ -124,30 +182,42 @@ describe('updateProcessEnv(launchEnv)', function () {
-        expect(shouldGetEnvFromShell('/usr/local/bin/zsh')).toBe(true)
-        expect(shouldGetEnvFromShell('/bin/fish')).toBe(true)
-        expect(shouldGetEnvFromShell('/usr/local/bin/fish')).toBe(true)
+        expect(shouldGetEnvFromShell({SHELL: '/bin/sh'})).toBe(true)
```
Do we need to test all these now that we have changed the logic? It seems like this test could be combined with the one below it.

joefitzgerald replied:
These were removed in a prior change and that resulted in a regression for zsh. While these may be perceived as redundant, I think they test for the vast majority of shells used by users, and so we should leave the assertions in there to guarantee no regressions occur.

nathansobo commented on spec/update-process-env-spec.js:
```diff
@@ -57,31 +60,60 @@ describe('updateProcessEnv(launchEnv)', function () {
       })
     })
+
+    it('allows ATOM_SUPPRESS_ENV_PATCHING to be preserved if set', function () {
```
Why does preserving this variable matter? It seems like we only ever look for it in the incoming environment to determine whether to mutate `process.env`. Does its presence in `process.env` itself ever matter?

nathansobo replied:
I see now, if we *ever* launched from a shell, we'll never shell out for the environment going forward. Makes sense. Disregard above comment.

joefitzgerald commented:
> If I'm reading this correctly, it seems like the name of the variable ATOM_SUPPRESS_ENV_PATCHING is a bit misleading. We will update the process.env of an existing instance of Atom regardless of this flag. The flag indicates whether we should attempt to invoke the user's shell to retrieve the environment... am I right about that?

There are two separate behaviors going on in update-environment:

1. We are potentially patching the environment by launching a new, interactive, login shell and using that environment - if Atom was *not* launched from the shell
1. We are allowing multiple invocations of `atom` from the shell to update the environment with new environment values; we are also allowing for the scenario where you open atom to a directory, and then subsequently open that same directory via `atom ./thedir` and allowing the environment to be updated with the environment from the shell

So, there's patching, and there's updating of the environment. They are different. Suppressing the patching should not suppress the updating. Maybe we need new verbs ;)

joefitzgerald commented:
Also worth mentioning: it should be possible for a user to set `ATOM_SUPPRESS_ENV_PATCHING` in their environment used by `launchd` to run apps launched by Dock / Finder / Spotlight to completely suppress environment patching. I believe this will be the minority of users, but I suspect some users will want to do this. It would be nice to provide some other configuration file that could be used to suppress this behavior - possibly residing in `ATOM_HOME` as a future extension / in a subsequent PR.

joefitzgerald committed:
Add Spec To Validate Updating An Existing Env Var

- :art: Update spec description to more accurately reflect the intent

joefitzgerald committed:
Combine Check

- The additional (!process.env[key] && envToAssign[key]) check allows “preserved” variables to be set for the first time if they are currently unset

nathansobo commented on src/update-process-env.js:
```diff
@@ -34,7 +31,7 @@ function updateProcessEnv (launchEnv) {
 
     for (let key in envToAssign) {
-      if (!ENVIRONMENT_VARIABLES_TO_PRESERVE.has(key)) {
+      if (!ENVIRONMENT_VARIABLES_TO_PRESERVE.has(key) || (!process.env[key] && envToAssign[key])) {
```
Why would `envToAssign[key]` be falsy? You're thinking like a blank string environment variable? The key definitely exists because we're looping over keys in the line above.

joefitzgerald replied:
Suppose you launch atom from Finder. `ATOM_SUPPRESS_ENV_PATCHING` is not set. It is also a "preserved" environment variable. If you were to subsequently launch Atom via the shell, it would never be set, and would not be preserved in future invocations.

nathansobo replied:
Right, that explains the `!process.env[key]` part of the expression but not the `&& envToAssign[key]`. It's minor though. I'll just merge anyway.

steakknife replied:
For posterity: That last part is unnecessary because `key` is always unique.

nathansobo commented:
> Maybe we need new verbs ;)

We definitely need new verbs. Patching and updating are the same thing. We need a clearer description of what we're suppressing that focuses on how we obtain the thing we're using to update the current environment... `DISABLE_SHELLING_OUT_FOR_ENVIRONMENT`? I'm not happy with the clarity of the existing naming.

joefitzgerald committed:
ATOM_SUPPRESS_ENV_PATCHING > ATOM_DISABLE_SHELLING_OUT_FOR_ENVIRONMENT

joefitzgerald commented:
Renamed `ATOM_SUPPRESS_ENV_PATCHING` to `ATOM_DISABLE_SHELLING_OUT_FOR_ENVIRONMENT`.

nathansobo commented:
Thanks @joefitzgerald. This definitely feels like a more robust approach.
